twap/main.go

```diff
@@ -0,0 +1,219 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/blocky/as-demo/as"
+	"github.com/blocky/as-demo/price"
+)
+
+type ArgsIterate struct {
+	TokenAddress string          `json:"token_address"`
+	ChainID      string          `json:"chain_id"`
+	NumSamples   int             `json:"num_samples"`
+	EAttest      json.RawMessage `json:"eAttest"`
+	TAttest      json.RawMessage `json:"tAttest"`
+	Whitelist    json.RawMessage `json:"whitelist"`
+}
+
+type ArgsTWAP struct {
+	EAttest   json.RawMessage `json:"eAttest"`
+	TAttest   json.RawMessage `json:"tAttest"`
+	Whitelist json.RawMessage `json:"whitelist"`
+}
+
+type Result struct {
+	Success bool
+	Error   string
+}
+
+type PriceSamples []price.Price
+
+type ResultPriceSamples struct {
+	Result       Result
+	PriceSamples PriceSamples
+}
+
+type ResultTWAP struct {
+	Result Result
+	TWAP   float64
+}
+
+func extractPriceSamples(
+	eAttest,
+	tAttest,
+	whitelist json.RawMessage,
+) (
+	PriceSamples,
+	error,
+) {
+	// bootstrap with empty samples if we don't have a transitive attestation
+	if tAttest == nil {
+		return PriceSamples{}, nil
+	}
+
+	verifyOut, err := as.VerifyAttestation(
+		as.HostVerifyAttestationInput{
+			EnclaveAttestedKey:    eAttest,
+			TransitiveAttestation: tAttest,
+			AcceptableMeasures:    whitelist,
+		},
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not verify previous attestation: %w", err)
+	}
+
+	var fixedRep [][]byte
+	err = json.Unmarshal(verifyOut.RawClaims, &fixedRep)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal previous claims: %w", err)
+	}
+
+	prevResultData := fixedRep[3]
+	var prevResult ResultPriceSamples
+	err = json.Unmarshal(prevResultData, &prevResult)
+	switch {
+	case err != nil:
+		return nil, fmt.Errorf("could not unmarshal previous output: %w", err)
+	case !prevResult.Result.Success:
+		return nil, fmt.Errorf("previous run was an error: %w", err)
+	}
+
+	return prevResult.PriceSamples, nil
+
+}
+
+func getNewPriceSample(tokenAddress string, chainID string) (price.Price, error) {
+	req := as.HostHTTPRequestInput{
+		Method: "GET",
+		URL: fmt.Sprintf(
+			"https://app.steer.finance/api/token/price?tokenAddress=%s&chainId=%s",
+			tokenAddress,
+			chainID,
+		),
+	}
+	resp, err := as.HostFuncHTTPRequest(req)
+	if err != nil {
+		return price.Price{}, fmt.Errorf("making http request: %w", err)
+	}
+
+	steerData := struct {
+		Price float64 `json:"price"`
+	}{}
+
+	err = json.Unmarshal(resp.Body, &steerData)
+	if err != nil {
+		return price.Price{}, fmt.Errorf(
+			"unmarshaling coin gecko data: %w...%s",
```

Comment:
this error doesn't make sense if the data is coming from steer.

Reply:
udpating

Reply:
updating